sale_coupon_auto_refresh/models/sale_order.py

```diff
@@ -10,21 +12,17 @@ class SaleOrder(models.Model):
     auto_refresh_coupon = fields.Boolean(related="company_id.auto_refresh_coupon",)
 
     def _auto_refresh_coupons(self):
-        self.filtered(
-            lambda x: x.state in ("draft", "sent") and x.auto_refresh_coupon
-        ).recompute_coupon_lines()
-
-    @api.model_create_multi
-    def create(self, vals_list):
-        """Create or refresh coupon lines on create"""
-        orders = super().create(vals_list)
-        orders.with_context(skip_auto_refresh_coupons=True)._auto_refresh_coupons()
-        return orders
-
-    def write(self, vals):
-        """Create or refresh coupon lines after saving"""
```

Comment:
What if the customer changes? A coupon could not apply anymore...

A nice thing about this approach is that it doesn't rely on the order line writes, but in sale.order ones. So when you save your changes in a sale order view, it gets here (it updates order_line), but if you directly make a write on the line,  it won't (so you avoid a good amount of undesired calls).

Reply:
@chienandalu I'll add the `_auto_refresh_coupons` to the `sale.order` `write` method too.

However, there are lots of cases in which a `sale.order.line` record is updated via code, not UI, without calling the `sale.order` `write` itself, therefore it may be safe to leave the implementation also on line level.

Reply:
Can you point a case? I can think of website sales, but precisely those update the promotions automatically already...

Reply:
I recall some cases in timesheet management, if I'm not mistaken; however, I think it's safer to manage both models' updates anyway.